nav2_util/include/nav2_util/geometry_utils.hpp

```diff
@@ -179,6 +181,30 @@ inline double calculate_path_length(const nav_msgs::msg::Path & path, size_t sta
   return path_length;
 }
 
+/**
+ * @brief Find the index of the first goal in `PENDING` status matching the
+ * given target pose.
+ * @param waypoint_statuses List of waypoint statuses to search through.
+ * @param goal Target pose to match against waypoint goals.
+ * @return Index of the first matching goal in PENDING status, -1 if not found.
+ */
+inline int find_goal_in_waypoint_statuses(
```

Comment:
This needs unit testing :-)

Reply:
I think this also needs to be renamed, since its not just finding the goal, its finding the next pending goal. How about 'find_next_matching_goal_in_waypoint_statuses'?